core/base/types.hpp

```diff
@@ -153,6 +153,188 @@ using default_precision = double;
+ * number of exponent bits, and thus preserving the range of representable
+ * values) or using one of the hardware or software supported conversions
+ * between IEEE formats, such as double to float or float to half (reducing both
+ * the number of exponent, as well as significand bits, and thus decreasing the
+ * range of representable values).
+ *
+ * The precision_reduction class encodes the lower precision format relative to
+ * the base precision used and the algorithm in question. The encoding is
+ * done by specifying the amount of range non-preserving conversions and the
+ * amount of range preserving conversions that should be done on the base
+ * precision to obtain the lower precision format. For example, starting with a
+ * double precision value (11 exp, 52 sig. bits), the encoding specifying 1
+ * non-preserving conversion and 1 preserving conversion would first use a
+ * hardware-supported non-preserving conversion to obtain a single precision
+ * value (8 exp, 23 sig. bits), followed by a preserving bit truncation to
+ * obtain a value with 8 exponent and 7 significand bits. Note that
+ * non-preserving conversion are always done first, as preserving conversions
+ * usually result in datatypes that are not supported by builtin conversions
+ * (thus, it is generally not possible to apply a non-preserving conversion to
+ * the result of a preserving conversion).
+ *
+ * If the specified conversion is not supported by the algorithm, it will most
+ * likely fall back to using full precision for storing the data. Refer to the
+ * documentation of specific algorithms using this class for details about such
+ * special cases.
+ */
+class precision_reduction {
+public:
+    /**
+     * The underlying datatype used to store the encoding.
+     */
+    using storage_type = uint8;
+
+private:
+    static constexpr auto nonpreserving_bits = 4u;
+    static constexpr auto preserving_bits =
+        byte_size * sizeof(storage_type) - nonpreserving_bits;
+    static constexpr auto nonpreserving_mask =
+        storage_type{(0x1 << nonpreserving_bits) - 1};
+    static constexpr auto preserving_mask =
+        storage_type{(0x1 << preserving_bits) - 1} << nonpreserving_bits;
+
+public:
+    /**
+     * Creates a default precision_reduction encoding.
+     *
+     * This encoding represents the case where no conversions are performed.
+     */
+    GKO_ATTRIBUTES constexpr precision_reduction() noexcept : data_{0x0} {}
+
+    /**
+     * Creates a precision_reduction encoding with the specified number of
+     * conversions.
+     *
+     * @param preserving  the number of range preserving conversion
+     * @param nonpreserving  the number of range non-preserving conversions
+     */
+    GKO_ATTRIBUTES constexpr precision_reduction(
+        storage_type preserving, storage_type nonpreserving) noexcept
+        : data_((preserving << nonpreserving_bits) | nonpreserving)
+    {}
```

Comment:
I would probably just ignore it for now, to keep the implementation as lightweight as possible.
Any other solution needs additional instructions to handle such cases, and the handling won't be perfect:

- we cannot throw an exception since that does not work on GPUs
- just setting it to "something" may not be what the user want

Reply:
The only thing we could do for now is put an assertion here which will cause the program to fail at runtime in such cases, if compiled in debug mode.

Reply:
I'll see what I can do, it's probably not going to look very nice because of the C++11 `constexpr`  restrictions, which means I have to introduce an assert statement, without actually writing a statement. I can probably use a `,` operator on something to insert it as part of the expression.

Reply:
Just leave it like it is, otherwise others won't understand what is done here if you use the `,` operator.